File: bins/astra-sim/src/battery.rs
//! Async battery simulator — port of `simulator/battery.py`.
//!
//! Speaks the CT002 UDP protocol: builds requests with the current
//! output power, sends them to the configured CT002 emulator, parses
//! the response, and adjusts its own target to drive the reported
//! grid reading toward zero.

use std::sync::Arc;
use std::time::Duration;

use parking_lot::Mutex;
use rand::Rng;
use tokio::net::UdpSocket;

use crate::protocol;

/// Snapshot of one battery's externally-visible state.
#[derive(Debug, Clone, serde::Serialize)]
pub struct BatteryState {
    pub mac: String,
    pub phase: String,
    pub power: i64,
    pub target: i64,
    pub applied_target: i64,
    pub power_update_delay_ticks: i64,
    pub soc: f64,
    pub max_charge: i64,
    pub max_discharge: i64,
}

/// Battery simulator. Public methods are thread-safe; the physics state
/// is behind a `parking_lot::Mutex` so the HTTP control endpoints and
/// the `step` loop can race without unsoundness.
pub struct BatterySimulator {
    pub mac: String,
    pub phase: char,
    pub ct_mac: String,
    pub ct_host: String,
    pub ct_port: u16,
    pub meter_dev_type: String,
    pub ct_dev_type: String,
    pub min_power_threshold: f64,
    pub ramp_rate: f64,
    pub poll_interval: f64,
    pub startup_delay: f64,
    pub inspection_count: u32,
    pub time_scale: f64,
    state: Mutex<State>,
}

struct State {
    current_power: f64,
    soc: f64,
    capacity_wh: f64,
    target_power: f64,
    requested_target: f64,
    request_count: u32,
    startup_elapsed: f64,
    step_index: i64,
    pending_targets: Vec<(i64, f64)>,
    max_charge_power: i64,
    max_discharge_power: i64,
    power_update_delay_ticks: i64,
}

#[derive(Clone)]
pub struct BatteryConfig {
    pub mac: String,
    pub phase: char,
    pub ct_mac: String,
    pub ct_host: String,
    pub ct_port: u16,
    pub meter_dev_type: String,
    pub ct_dev_type: String,
    pub max_charge_power: i64,
    pub max_discharge_power: i64,
    pub capacity_wh: f64,
    pub initial_soc: f64,
    pub ramp_rate: f64,
    pub poll_interval: f64,
    pub min_power_threshold: f64,
    pub startup_delay: f64,
    pub inspection_count: u32,
    pub time_scale: f64,
    pub power_update_delay_ticks: i64,
}

impl BatterySimulator {
    pub fn new(cfg: BatteryConfig) -> anyhow::Result<Arc<Self>> {
        if protocol::phase_field_index(cfg.phase).is_none() {
            anyhow::bail!("Invalid phase {:?}", cfg.phase);
        }
        Ok(Arc::new(Self {
            mac: cfg.mac.to_uppercase(),
            phase: cfg.phase,
            ct_mac: cfg.ct_mac,
            ct_host: cfg.ct_host,
            ct_port: cfg.ct_port,
            meter_dev_type: cfg.meter_dev_type,
            ct_dev_type: cfg.ct_dev_type,
            min_power_threshold: cfg.min_power_threshold,
            ramp_rate: cfg.ramp_rate,
            poll_interval: cfg.poll_interval,
            startup_delay: cfg.startup_delay.max(0.0),
            inspection_count: cfg.inspection_count,
            time_scale: cfg.time_scale.max(0.1),
            state: Mutex::new(State {
                current_power: 0.0,
                soc: cfg.initial_soc.clamp(0.0, 1.0),
                capacity_wh: cfg.capacity_wh,
                target_power: 0.0,
                requested_target: 0.0,
                request_count: 0,
                startup_elapsed: 0.0,
                step_index: 0,
                pending_targets: Vec::new(),
                max_charge_power: cfg.max_charge_power,
                max_discharge_power: cfg.max_discharge_power,
                power_update_delay_ticks: cfg.power_update_delay_ticks.max(0),
            }),
        }))
    }

    pub fn current_power(&self) -> f64 {
        self.state.lock().current_power
    }

    pub fn snapshot(&self) -> BatteryState {
        let s = self.state.lock();
        BatteryState {
            mac: self.mac.clone(),
            phase: self.phase.to_string(),
            power: s.current_power.round() as i64,
            target: s.requested_target.round() as i64,
            applied_target: s.target_power.round() as i64,
            power_update_delay_ticks: s.power_update_delay_ticks,
            soc: (s.soc * 10000.0).round() / 10000.0,
            max_charge: s.max_charge_power,
            max_discharge: s.max_discharge_power,
        }
    }

    pub fn set_soc(&self, soc: f64) {
        self.state.lock().soc = soc.clamp(0.0, 1.0);
    }

    pub fn set_max_charge(&self, w: i64) {
        self.state.lock().max_charge_power = w.max(0);
    }

    pub fn set_max_discharge(&self, w: i64) {
        self.state.lock().max_discharge_power = w.max(0);
    }

    /// Apply a CT-derived target: with `power_update_delay_ticks == 0` it
    /// takes effect immediately; otherwise it is queued for `N` ticks
    /// later. Exposed for tests that drive the state machine directly.
    pub fn apply_ct_derived_target(&self, new_target: f64) {
        let mut s = self.state.lock();
        s.requested_target = new_target;
        if s.power_update_delay_ticks <= 0 {
            s.target_power = new_target;
            return;
        }
        let apply_at = s.step_index + s.power_update_delay_ticks;
        s.pending_targets.push((apply_at, new_target));
    }

    fn drain_pending(&self) {
        let mut s = self.state.lock();
        if s.power_update_delay_ticks <= 0 {
            return;
        }
        let cur = s.step_index;
        let mut remaining = Vec::new();
        // Walk the pending list once. Apply any whose `apply_at <= step_index`
        // directly to `target_power` (last in iteration order wins, mirroring
        // Python's `_drain_pending_power_targets`). Keep the rest.
        for (apply_at, target) in s.pending_targets.drain(..).collect::<Vec<_>>() {
            if apply_at <= cur {
                s.target_power = target;
            } else {
                remaining.push((apply_at, target));
            }
        }
        s.pending_targets = remaining;
    }

    fn update_power(&self, dt: f64) {
        let mut s = self.state.lock();
        let mut target = s.target_power;
        if target.abs() < self.min_power_threshold {
            target = 0.0;
        }
        if s.soc >= 1.0 && target < 0.0 {
            target = 0.0;
        }
        if s.soc <= 0.0 && target > 0.0 {
            target = 0.0;
        }
        // Startup delay.
        if self.startup_delay > 0.0 {
            let idle = s.current_power.abs() < self.min_power_threshold;
            let want_power = target.abs() >= self.min_power_threshold;
            if idle && want_power {
                s.startup_elapsed += dt;
                if s.startup_elapsed < self.startup_delay {
                    return;
                }
            } else {
                s.startup_elapsed = 0.0;
            }
        }
        let diff = target - s.current_power;
        let max_step = self.ramp_rate * dt;
        let step = if diff.abs() > max_step {
            if diff > 0.0 {
                max_step
            } else {
                -max_step
            }
        } else {
            diff
        };
        s.current_power += step;
        let max_d = s.max_discharge_power as f64;
        let max_c = -(s.max_charge_power as f64);
        s.current_power = s.current_power.clamp(max_c, max_d);
    }

    fn update_soc(&self, dt: f64) {
        let mut s = self.state.lock();
        if s.capacity_wh <= 0.0 {
            return;
        }
        let energy_wh = s.current_power * (dt / 3600.0);
        s.soc -= energy_wh / s.capacity_wh;
        s.soc = s.soc.clamp(0.0, 1.0);
    }

    /// Send a single CT002 request and apply the response's grid reading
    /// (sum of phases A+B+C) as the new target. Returns the parsed
    /// response fields or `None` if the request timed out / decoded badly.
    async fn send_request(&self) -> Option<Vec<String>> {
        let (req_count, current) = {
            let s = self.state.lock();
            (s.request_count, s.current_power)
        };
        let phase_field = if req_count < self.inspection_count {
            "0".to_string()
        } else {
            self.phase.to_string()
        };
        let fields = vec![
            self.meter_dev_type.clone(),
            self.mac.clone(),
            self.ct_dev_type.clone(),
            self.ct_mac.clone(),
            phase_field.clone(),
            current.round().to_string(),
        ];
        let payload = protocol::build_payload(&fields);

        let sock = match UdpSocket::bind("0.0.0.0:0").await {
            Ok(s) => s,
            Err(e) => {
                tracing::debug!("Battery {}: bind: {e}", self.mac);
                return None;
            }
        };
        let dst = format!("{}:{}", self.ct_host, self.ct_port);
        if let Err(e) = sock.send_to(&payload, &dst).await {
            tracing::debug!("Battery {}: send: {e}", self.mac);
            return None;
        }
        let mut buf = vec![0u8; 2048];
        let n = match tokio::time::timeout(Duration::from_secs(2), sock.recv_from(&mut buf)).await {
            Ok(Ok((n, _))) => n,
            _ => {
                tracing::debug!("Battery {}: recv timeout", self.mac);
                return None;
            }
        };
        self.state.lock().request_count += 1;
        let (parsed, err) = protocol::parse_message(&buf[..n]);
        if let Some(e) = err {
            tracing::debug!("Battery {}: bad response: {e}", self.mac);
            return None;
        }
        let fields = parsed?;
        // Extract grid reading: sum of phase A/B/C power fields (4,5,6).
        if phase_field != "0" {
            let pa: i64 = fields.get(4).and_then(|s| s.parse().ok()).unwrap_or(0);
            let pb: i64 = fields.get(5).and_then(|s| s.parse().ok()).unwrap_or(0);
            let pc: i64 = fields.get(6).and_then(|s| s.parse().ok()).unwrap_or(0);
            let grid_reading = (pa + pb + pc) as f64;
            self.apply_ct_derived_target(current + grid_reading);
        }
        Some(fields)
    }

    /// Execute one simulator iteration with explicit `dt`. Does NOT
    /// sleep — designed for deterministic testing.
    pub async fn step(self: &Arc<Self>, dt: f64) -> Option<Vec<String>> {
        {
            let mut s = self.state.lock();
            s.step_index += 1;
        }
        self.drain_pending();
        self.update_power(dt);
        self.update_soc(dt);
        self.send_request().await
    }

    /// Long-running poll loop with random jitter.
    pub async fn run(self: Arc<Self>, cancel: tokio_util_local::CancelFlag) {
        tracing::info!(
            "Battery {} started (phase={}, soc={:.0}%)",
            self.mac,
            self.phase,
            self.state.lock().soc * 100.0
        );
        let mut last = std::time::Instant::now();
        loop {
            if cancel.is_cancelled() {
                return;
            }
            let now = std::time::Instant::now();
            let dt = now.duration_since(last).as_secs_f64() * self.time_scale;
            last = now;
            let _ = self.step(dt).await;
            let jitter = rand::thread_rng().gen_range(-0.5..=0.5);
            let sleep = ((self.poll_interval + jitter) / self.time_scale).max(0.05);
            tokio::time::sleep(Duration::from_secs_f64(sleep)).await;
        }
    }
}

// Tiny inline cancellation helper to avoid pulling tokio_util just for this.
pub mod tokio_util_local {
    use std::sync::atomic::{AtomicBool, Ordering};
    use std::sync::Arc;

    #[derive(Clone, Default)]
    pub struct CancelFlag(Arc<AtomicBool>);

    impl CancelFlag {
        pub fn new() -> Self {
            Self::default()
        }
        pub fn cancel(&self) {
            self.0.store(true, Ordering::SeqCst);
        }
        pub fn is_cancelled(&self) -> bool {
            self.0.load(Ordering::SeqCst)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cfg() -> BatteryConfig {
        BatteryConfig {
            mac: "AABBCCDDEEFF".into(),
            phase: 'A',
            ct_mac: "112233445566".into(),
            ct_host: "127.0.0.1".into(),
            ct_port: 0,
            meter_dev_type: "HMG-50".into(),
            ct_dev_type: "HME-4".into(),
            max_charge_power: 800,
            max_discharge_power: 800,
            capacity_wh: 2560.0,
            initial_soc: 0.5,
            ramp_rate: 200.0,
            poll_interval: 1.0,
            min_power_threshold: 20.0,
            startup_delay: 0.0,
            inspection_count: 0,
            time_scale: 1.0,
            power_update_delay_ticks: 0,
        }
    }

    #[test]
    fn soc_clamps_at_extremes() {
        let b = BatterySimulator::new(cfg()).unwrap();
        b.set_soc(5.0);
        assert_eq!(b.snapshot().soc, 1.0);
        b.set_soc(-1.0);
        assert_eq!(b.snapshot().soc, 0.0);
    }

    #[tokio::test]
    async fn ramps_toward_target_within_max_step() {
        let b = BatterySimulator::new(cfg()).unwrap();
        b.apply_ct_derived_target(500.0);
        // dt=1s, ramp_rate=200 → expect ~200W after one step (no UDP roundtrip).
        b.update_power(1.0);
        let snap = b.snapshot();
        assert!(snap.power == 200, "got {}", snap.power);
    }

    /// Ports `test_power_update_immediate_when_delay_zero`.
    #[test]
    fn power_update_immediate_when_delay_zero() {
        let b = BatterySimulator::new(cfg()).unwrap();
        b.state.lock().current_power = 100.0;
        b.apply_ct_derived_target(250.0);
        assert_eq!(b.state.lock().target_power, 250.0);
    }

    /// Ports `test_power_update_delayed_by_n_ticks`. With delay=2, the CT
    /// target observed at tick T applies at tick T+2.
    #[test]
    fn power_update_delayed_by_n_ticks() {
        let c = BatteryConfig {
            power_update_delay_ticks: 2,
            startup_delay: 0.0,
            min_power_threshold: 0.0,
            ramp_rate: 1e9,
            inspection_count: 0,
            ..cfg()
        };
        let b = BatterySimulator::new(c).unwrap();
        b.state.lock().current_power = 100.0;
        // Each step bumps step_index by 1 and then drains pending. Simulate
        // the Python test by enqueuing the target *before* the tick that
        // owns it, i.e. in step 1 enqueue (applies in step 3).
        b.state.lock().step_index = 1;
        b.apply_ct_derived_target(250.0);
        // step 2 — not yet
        b.state.lock().step_index = 2;
        b.drain_pending();
        assert_eq!(b.state.lock().target_power, 0.0);
        b.state.lock().step_index = 3;
        b.drain_pending();
        assert_eq!(b.state.lock().target_power, 250.0);
    }
}

File: src/astrameter/simulator/battery.py
"""Async Marstek battery simulator.

Speaks the CT002 UDP protocol, sends periodic requests to the CT002
emulator, receives per-phase power targets, and adjusts its simulated
output accordingly.
"""

from __future__ import annotations

import asyncio
import logging
import random
import time

from astrameter.ct002.balancer import device_capabilities

from . import protocol
from .b2500_steering import B2500SteeringController
from .firmware_steering import FirmwareSteeringController
from .venus_d_steering import VenusDSteeringController

logger = logging.getLogger("astra_sim.battery")


class BatterySimulator:
    def __init__(
        self,
        mac: str,
        phase: str,
        ct_mac: str,
        ct_host: str = "127.0.0.1",
        ct_port: int = 12345,
        meter_dev_type: str = "HMG-50",
        ct_dev_type: str = "HME-4",
        max_charge_power: int = 800,
        max_discharge_power: int = 800,
        capacity_wh: float = 2560.0,
        initial_soc: float = 0.5,
        ramp_rate: float = 200.0,
        poll_interval: float = 1.0,
        min_power_threshold: float = 20.0,
        startup_delay: float = 2.0,
        inspection_count: int = 1,
        time_scale: float = 1.0,
        power_update_delay_ticks: int = 0,
        max_dc_input: int = 0,
        dc_input_power: float = 0.0,
        participates: bool = True,
    ) -> None:
        if phase not in protocol.PHASE_FIELD_INDEX:
            raise ValueError(
                f"Invalid phase {phase!r}, must be one of "
                f"{list(protocol.PHASE_FIELD_INDEX)}"
            )

        self.mac = mac.upper()
        self.phase = phase
        self.ct_mac = ct_mac
        self.ct_host = ct_host
        self.ct_port = ct_port
        self.meter_dev_type = meter_dev_type
        self.ct_dev_type = ct_dev_type
        self.max_charge_power = max_charge_power
        self.max_discharge_power = max_discharge_power
        self.capacity_wh = capacity_wh
        self.ramp_rate = ramp_rate
        self.poll_interval = poll_interval
        self.min_power_threshold = min_power_threshold
        self.startup_delay = max(0.0, startup_delay)
        self.inspection_count = inspection_count
        self.time_scale = max(0.1, time_scale)
        self.power_update_delay_ticks = max(0, int(power_update_delay_ticks))
        self.max_dc_input = max(0, int(max_dc_input))
        self.participates = participates

        self._current_power: float = 0.0
        self._soc: float = max(0.0, min(1.0, initial_soc))
        self._target_power: float = 0.0
        self._requested_target: float = 0.0
        self._request_count: int = 0
        self._last_update: float = time.monotonic()
        self._startup_elapsed: float = 0.0
        self._step_index: int = 0
        self._pending_power_targets: list[tuple[int, float]] = []
        self._dc_input_power: float = 0.0
        self.dc_input_power = dc_input_power  # reuse setter clamp

        # Self-consumption control law. Most Marstek batteries (Venus class) run
        # the firmware ramp controller on the grid value read back from the CT;
        # ``hi``/``lo`` are the charge / discharge limits in its convention
        # (setpoint positive = charge, negative = discharge).
        self._steering = FirmwareSteeringController()
        self._steer_hi = float(self.max_charge_power)
        self._steer_lo = -float(self.max_discharge_power)

        # The B2500 family (HMA/HMJ/HMK) is DC-coupled with no built-in inverter
        # and no AC input: it steers its DC output into an external microinverter
        # with a different integer hysteresis law instead of the Venus ramp. The
        # unit has two DC output channels, each its own regulator running every
        # cycle; the grid-derived demand is split between them, so the combined
        # output slews at twice a single channel's rate (~34 vs ~17 W/cycle).
        caps = device_capabilities(self.meter_dev_type)
        self._is_dc_output = (
            caps.has_dc_input
            and not caps.has_builtin_inverter
            and not caps.has_ac_input
        )
        # Two channels, each capped at half the unit's discharge envelope.
        _ch_max = max(1, self.max_discharge_power // 2)
        self._b2500_channels = (
            [
                B2500SteeringController(max_output=_ch_max),
                B2500SteeringController(max_output=_ch_max),
            ]
            if self._is_dc_output
            else []
        )

        # Venus D (VNSD-0) is AC-coupled like the rest of the Venus class but
        # runs a different self-consumption loop: an integer proportional
        # integrator rather than the float ramp (see :mod:`venus_d_steering`).
        # Its setpoint convention is the opposite of the ramp controller's —
        # positive = discharge, negative = charge — so ``hi`` is the discharge
        # limit and ``lo`` the (negative) charge limit, and the simulator target
        # is the setpoint *unnegated*.
        self._is_venus_d = (
            self.meter_dev_type.upper().startswith("VNSD") and not self._is_dc_output
        )
        self._venus_d_steering = (
            VenusDSteeringController() if self._is_venus_d else None
        )

    # -- public read-only properties ---------------------------------------

    @property
    def current_power(self) -> float:
        return self._current_power

    @current_power.setter
    def current_power(self, value: float) -> None:
        self._current_power = value

    @property
    def soc(self) -> float:
        return self._soc

    @soc.setter
    def soc(self, value: float) -> None:
        self._soc = max(0.0, min(1.0, value))

    @property
    def target_power(self) -> float:
        return self._target_power

    @property
    def dc_input_power(self) -> float:
        return self._dc_input_power

    @dc_input_power.setter
    def dc_input_power(self, value: float) -> None:
        self._dc_input_power = max(0.0, min(float(self.max_dc_input), float(value)))

    def _apply_ct_derived_target(self, new_target: float) -> None:
        """Record CT request immediately; apply to physics after *power_update_delay_ticks*."""
        self._requested_target = new_target
        if self.power_update_delay_ticks <= 0:
            self._target_power = new_target
            return
        apply_at = self._step_index + self.power_update_delay_ticks
        self._pending_power_targets.append((apply_at, new_target))

    def _drain_pending_power_targets(self) -> None:
        if self.power_update_delay_ticks <= 0:
            return
        remaining: list[tuple[int, float]] = []
        for apply_at, target in self._pending_power_targets:
            if apply_at <= self._step_index:
                self._target_power = target
            else:
                remaining.append((apply_at, target))
        self._pending_power_targets = remaining

    # -- physics -----------------------------------------------------------

    def _update_power(self, dt: float) -> None:
        target = self._target_power

        if abs(target) < self.min_power_threshold:
            target = 0.0

        # SOC saturation
        if self._soc >= 1.0 and target < 0:
            target = 0.0
        if self._soc <= 0.0 and target > 0:
            target = 0.0

        # Startup delay: when resuming from idle the real inverter needs
        # a few seconds before it begins ramping.  During this window the
        # battery stays at ~0 W, which is the behaviour that previously
        # triggered false saturation detection.
        if self.startup_delay > 0:
            idle = abs(self._current_power) < self.min_power_threshold
            want_power = abs(target) >= self.min_power_threshold
            if idle and want_power:
                self._startup_elapsed += dt
                if self._startup_elapsed < self.startup_delay:
                    self._apply_dc_passthrough()
                    return  # stay at current (near-zero) power
            else:
                self._startup_elapsed = 0.0

        # Ramp toward target
        diff = target - self._current_power
        max_step = self.ramp_rate * dt
        if abs(diff) > max_step:
            diff = max_step if diff > 0 else -max_step
        self._current_power += diff

        # Clamp to limits
        self._current_power = max(
            -self.max_charge_power,
            min(self.max_discharge_power, self._current_power),
        )

        # When SoC is saturated and DC input is present, the inverter
        # passes the unabsorbed PV through to AC even if the AC target
        # asks for charging.  Mirrors Marstek Venus D behaviour.
        self._apply_dc_passthrough()

    def _apply_dc_passthrough(self) -> None:
        if self._soc < 1.0 or self._dc_input_power <= 0:
            return
        # Push at least the DC input through to AC as positive output.
        self._current_power = max(self._current_power, self._dc_input_power)

    def _update_soc(self, dt: float) -> None:
        if self.capacity_wh <= 0:
            return
        # AC energy first (positive current_power drains, negative charges)
        energy_wh = self._current_power * (dt / 3600.0)
        self._soc -= energy_wh / self.capacity_wh
        # DC input charges the cells in parallel (when not already full).
        if self._dc_input_power > 0:
            dc_energy_wh = self._dc_input_power * (dt / 3600.0)
            self._soc += dc_energy_wh / self.capacity_wh
        self._soc = max(0.0, min(1.0, self._soc))

    # -- protocol ----------------------------------------------------------

    def _request_fields(self) -> list[str]:
        """Build the CT002 request fields for this poll."""
        phase_field = "0" if self._request_count < self.inspection_count else self.phase
        fields = [
            self.meter_dev_type,
            self.mac,
            self.ct_dev_type,
            self.ct_mac,
            phase_field,
            str(round(self._current_power)),
        ]
        if not self.participates:
            # Opt out of CT aggregation via the optional 7th "participate"
            # field. Participating batteries omit it (matches Venus, which
            # sends only 6 fields).
            fields.append("0")
        return fields

    async def _send_request(self) -> list[str] | None:
        request_fields = self._request_fields()
        phase_field = request_fields[4]
        payload = protocol.build_payload(request_fields)

        loop = asyncio.get_running_loop()
        transport = None
        try:
            transport, proto = await asyncio.wait_for(
                loop.create_datagram_endpoint(
                    lambda: _UDPClient(),
                    remote_addr=(self.ct_host, self.ct_port),
                ),
                timeout=2.0,
            )
            transport.sendto(payload)
            data = await asyncio.wait_for(proto.received, timeout=2.0)
        except (TimeoutError, OSError) as exc:
            logger.debug("Battery %s: request failed: %s", self.mac, exc)
            return None
        finally:
            if transport is not None:
                transport.close()

        self._request_count += 1

        response_fields, err = protocol.parse_message(data)
        if err:
            logger.debug("Battery %s: bad response: %s", self.mac, err)
            return None

        # Hand parsed response off to the deterministic helper so it can
        # also be unit-tested without UDP I/O.
        if response_fields and phase_field != "0":
            self._handle_ct_response(response_fields)

        return response_fields

    def _handle_ct_response(self, response_fields: list[str]) -> None:
        """Derive the new AC target from the grid value read back from the CT.

        The grid value (sum of the per-phase power fields, positive = importing)
        is fed to this battery's steering controller. Venus-class batteries run
        :class:`FirmwareSteeringController` (a ramp law with input-conditioning
        gates), whose sign is the inverse of the simulator's (setpoint positive =
        charge), so the simulator target is the negated setpoint. A DC-coupled
        B2500 instead runs :class:`B2500SteeringController` on its DC output (see
        :meth:`_steer_b2500_output`). A Venus D (VNSD-0) runs
        :class:`VenusDSteeringController`, an integer integrator whose setpoint is
        already in the simulator's sign (positive = discharge), applied directly.

        Cross-battery share-split: a real battery divides the grid value by the
        number of batteries reported on its phase (the ``*_chrg_nb`` count), so
        several batteries on one phase each take their share rather than all
        chasing the full residual. This matters in relay mode / against a real
        CT; AstraMeter's active-control emulator distributes per-battery targets
        itself and reports a count of 1, so the split is a no-op there.
        """

        def field(idx: int) -> int:
            try:
                return int(response_fields[idx])
            except (IndexError, ValueError, TypeError):
                return 0

        grid_reading = field(4) + field(5) + field(6)

        if self._b2500_channels:
            self._steer_b2500_output(grid_reading)
            return

        if self._venus_d_steering is not None:
            # Venus D: integer integrator, positive setpoint = discharge. Its own
            # grid reading (used only for the per-step branch) tracks the CT
            # value in this closed loop. ±15 W deadband in combined (phase D)
            # mode, ±11 W otherwise.
            vd_setpoint = self._venus_d_steering.step(
                grid_reading,
                float(self.max_discharge_power),
                -float(self.max_charge_power),
                measured_grid=grid_reading,
                phase_count=2 if self.phase == "D" else 1,
            )
            self._apply_ct_derived_target(float(vd_setpoint))
            return

        # *_chrg_nb for this battery's phase (fields 9/10/11 → indices 8/9/10).
        phase_count = field(8 + "ABC".index(self.phase))

        setpoint = self._steering.step(
            grid_reading,
            self._steer_hi,
            self._steer_lo,
            device_count=phase_count,
            out=self._current_power,
        )
        # Controller: +charge → simulator: +discharge.
        self._apply_ct_derived_target(-setpoint)

    def _steer_b2500_output(self, grid_reading: int) -> None:
        """Steer a DC-coupled B2500's two output channels toward nulling the grid.

        The B2500 can only discharge its DC output into an external microinverter
        to offset the grid — it has no AC input and never charges from AC. The
        setpoint is **incremental** (current output plus 90% of the residual
        grid), so the loop integrates the grid toward zero rather than parking at
        a fraction of the load; it is floored at 0 (a surplus winds the output
        down to idle), capped at the discharge envelope, follows the B2500's
        integer hysteresis regulator rather than the Venus ramp, and is split
        evenly across the two output channels (each regulating against its own
        ~half of the measured output), so the combined output slews at ~twice a
        single channel's rate.
        """
        cur = max(0, round(self._current_power))
        target = cur + grid_reading * 9 // 10  # incremental: 90% of the residual
        target = max(0, min(target, self.max_discharge_power))
        # At full SoC the PV passes straight through to the output and cannot be
        # curtailed below the DC input (the pack is full, the PV has nowhere else
        # to go). The steering can't drive the output under that floor, so don't
        # let it try — otherwise it fights the passthrough override and the
        # output oscillates instead of settling at the PV level.
        if self._soc >= 1.0 and self._dc_input_power > 0:
            target = max(target, round(self._dc_input_power))
        per_channel = target // 2
        own = cur // 2  # each channel's ~half of the measured output
        out = sum(ch.regulate(per_channel, own) for ch in self._b2500_channels)
        self._apply_ct_derived_target(float(out))

    # -- main loop ---------------------------------------------------------

    async def step(self, dt: float | None = None) -> list[str] | None:
        """Execute one simulation iteration with explicit *dt*.

        When *dt* is ``None`` it defaults to :attr:`poll_interval`.
        Unlike :meth:`run`, this does **not** sleep or touch
        ``_last_update`` — it is designed for deterministic test
        stepping.
        """
        if dt is None:
            dt = self.poll_interval
        self._step_index += 1
        self._drain_pending_power_targets()
        self._update_power(dt)
        self._update_soc(dt)
        return await self._send_request()

    async def run(self) -> None:
        logger.info(
            "Battery %s started (phase=%s, soc=%.0f%%)",
            self.mac,
            self.phase,
            self._soc * 100,
        )
        self._last_update = time.monotonic()
        while True:
            now = time.monotonic()
            dt = (now - self._last_update) * self.time_scale
            self._last_update = now

            await self.step(dt)

            jitter = random.uniform(-0.5, 0.5)
            await asyncio.sleep(
                max(0.05, (self.poll_interval + jitter) / self.time_scale)
            )

    # -- serialisation -----------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "mac": self.mac,
            "phase": self.phase,
            "power": round(self._current_power),
            "target": round(self._requested_target),
            "applied_target": round(self._target_power),
            "power_update_delay_ticks": self.power_update_delay_ticks,
            "soc": round(self._soc, 4),
            "max_charge": self.max_charge_power,
            "max_discharge": self.max_discharge_power,
            "max_dc_input": self.max_dc_input,
            "dc_input": round(self._dc_input_power),
        }


class _UDPClient(asyncio.DatagramProtocol):
    """Minimal asyncio datagram protocol for a single request/response."""

    def __init__(self) -> None:
        self.received: asyncio.Future[bytes] = (
            asyncio.get_running_loop().create_future()
        )

    def datagram_received(self, data: bytes, addr: tuple[str, int]) -> None:
        if not self.received.done():
            self.received.set_result(data)

    def error_received(self, exc: Exception) -> None:
        if not self.received.done():
            self.received.set_exception(exc)
